pkg/operator/ceph/cluster/watcher.go

```diff
@@ -47,16 +50,24 @@ func newClientCluster(client client.Client, namespace string, context *clusterd.
 }
 
 // onK8sNodeAdd is triggered when a node is added in the Kubernetes cluster
-// func (c *clientCluster) onK8sNodeAdd(object runtime.Object) bool {
 func (c *clientCluster) onK8sNode(object runtime.Object) bool {
 	node, ok := object.(*v1.Node)
 	if !ok {
 		return false
 	}
-
+	// Check if the node have OSDs already present on it, reconcile if OSDs are not present
+	if nodesCheckedForReconcile.Contains(node.Name) {
+		return false
+	}
 	// Get CephCluster
 	cluster := c.getCephCluster()
 
+	if len(cluster.Spec.Storage.Nodes) == 0 && len(cluster.Spec.Storage.VolumeSources) == 0 && len(cluster.Spec.Storage.StorageClassDeviceSets) == 0 {
+		logger.Debugf("no nodes with storageClassDevicesets or volumeSources are specified, no OSD pods are going to be created")
+		nodesCheckedForReconcile.Add(node.Name) // adding node to nodesCheckedForReconcile Set
```

Comment:
nit: no need to have the comment, IMO it is already clear from the code
```suggestion
		nodesCheckedForReconcile.Add(node.Name)
```